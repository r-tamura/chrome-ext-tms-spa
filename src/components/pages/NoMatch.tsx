import * as React from "react";

export const NoMatch: React.FC = () => {
  return (
    <main className="main">
      <h1>Not Found 404</h1>
    </main>
  );
};
