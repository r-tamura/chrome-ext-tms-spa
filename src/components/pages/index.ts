export * from "./TimeCardPage";
export * from "./SignInPage";
export * from "./TransportExpensePage";
export * from "./NoMatch";
export * from "./App";
export * from "./Root";
