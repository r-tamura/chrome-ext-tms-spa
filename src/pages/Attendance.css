.attendance-holiday {
  color: rgba(255, 0, 0, .4);
  font-weight: var(--font-bold);
}
