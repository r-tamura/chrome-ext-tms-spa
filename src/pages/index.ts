export * from "./Attendance";
export * from "./SignInPage";
export * from "./TransportExpensePage";
export * from "./App";
export * from "./NoMatch";
