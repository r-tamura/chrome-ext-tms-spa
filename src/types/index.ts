export * from "./api_response";
export * from "./user";
export * from "./master";
export * from "./transexpenses";
export * from "./attendance";
