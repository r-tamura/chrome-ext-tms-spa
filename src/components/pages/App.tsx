import React from "react";
import { Helmet } from "react-helmet";
import { Route, Switch, Redirect } from "react-router-dom";
import { AppHeader, Nav } from "~/components/organisms";
import { EnsureLoggedInContainer } from "~/components/containers/EnsureLoggedInContainer";
import { TransportExpensePage, TimeCardPage, SignInPage, NoMatch } from ".";
import { AppGrid } from "~/components/organisms";

const PAGES_WITH_NAVBAR = ["/transportation", "/attendance"];

export function App() {
  return (
    <AppGrid>
      <Helmet>
        <link
          href="https://fonts.googleapis.com/css?family=M+PLUS+Rounded+1c"
          rel="stylesheet"
        ></link>
        <meta
          name="viewport"
          content="width=device-width, initial-scale=1.0, shrink-to-fit=no"
        />
      </Helmet>
      <AppHeader />
      <Route
        path={PAGES_WITH_NAVBAR}
        render={renderProps => {
          return <Nav path={renderProps.match.url} />;
        }}
      />
      <Switch>
        <Route
          exact={true}
          path="/"
          component={() => <Redirect to={"/signin"} />}
        />
        <Route exact={true} path="/signin" component={SignInPage} />
        <EnsureLoggedInContainer>
          {/* TODO: Dashboard画面作成までは交通費管理画面へリダイレクト */}
          <Route
            path="/dashboard"
            component={() => <Redirect to={"/transportation"} />}
          />
          <Route path="/transportation" component={TransportExpensePage} />
          <Route path="/attendance" component={TimeCardPage} />
        </EnsureLoggedInContainer>
        <Route component={NoMatch} />
      </Switch>
    </AppGrid>
  );
}
