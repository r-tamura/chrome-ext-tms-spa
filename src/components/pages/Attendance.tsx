import React from "react";
import styled from "styled-components";
import { Helmet } from "react-helmet";
import { SelectBox } from "~/components/SelectBox";
import { Button, ButtonGroup } from "~/components/atoms";
import { zerofill } from "~/helpers/common";
import { AttendanceMonthlyView } from "~/types";
import { MsgAttendance } from "~/helpers/_const";
import { useAttendance } from "~/stores/hooks";
import { MainContent } from "~/components/organisms";
import { BOLD } from "~/styles/font";

export const AttendancePage: React.FC = () => {
  const {
    projects,
    fetchAttendances,
    attendanceSettings,
    updateSettings,
    attendanceMonthly,
    saveAttendances,
    updateDaily,
    changeMonth,
    setMonthlyWithDefaults
  } = useAttendance();

  const { year, month } = attendanceMonthly;
  return (
    <main className="app-content-with-navbar">
      <Helmet>
        <title>Attendance | TMS</title>
      </Helmet>
      <MainContent>
        <h1>
          <Button id={"btn-prev-month"} onClick={onPrevMonthClick}>
            {"<"}
          </Button>
          {year} / {month}
          <Button id={"btn-next-month"} onClick={onNextMonthClick}>
            {">"}
          </Button>
        </h1>
        {renderMonthly(attendanceMonthly)}

        {/* 時間リスト */}
        <datalist id="timelist-am">
          <option>09:00</option>
          <option>09:30</option>
          <option>10:00</option>
          <option>10:30</option>
        </datalist>
        <datalist id="timelist-pm">
          <option>17:30</option>
          <option>18:00</option>
          <option>18:30</option>
          <option>19:00</option>
          <option>19:30</option>
          <option>19:30</option>
          <option>20:00</option>
          <option>20:30</option>
          <option>21:00</option>
          <option>21:30</option>
          <option>22:00</option>
          <option>22:30</option>
          <option>23:00</option>
          <option>23:30</option>
        </datalist>
      </MainContent>
    </main>
  );

  function renderMonthly({
    days,
    isFetching,
    hasApplied
  }: AttendanceMonthlyView) {
    if (isFetching) {
      // TODO: ロードインジケータを表示するように
      return <p>Now Loading...</p>;
    }

    if (days.length === 0) {
      return <p>勤怠データを取得できません</p>;
    }

    return (
      <div>
        <ButtonGroup>
          <Button id={"btn-set-default"} onClick={onSetDefaultClick}>
            set default
          </Button>
          <Button id={"btn-save"} onClick={onSaveClick}>
            Save
          </Button>
          <Button id={"btn-fetch"} onClick={fetchAttendances}>
            Reload
          </Button>
          {hasApplied ? (
            <Button id={"btn-submit"} title={"上長申請中"} disabled>
              Submit
            </Button>
          ) : (
            <Button id={"btn-submit"} title={"上長申請"} onClick={onSubmit}>
              Submit
            </Button>
          )}
        </ButtonGroup>
        <div>
          {/* <SelectBox
            value={attendanceSettings.projectId || ""}
            options={{
              items: projects,
              valueKey: "projectId",
              labelKey: "name"
            }}
            onChange={$select =>
              updateSettings({
                projectId: $select.selectedOptions[0].value
              })
            }
          /> */}
          select box
        </div>
        <table className="tms-table tms-table--bordered">
          <thead>
            <tr>
              <th>日</th>
              <th>出勤</th>
              <th>退勤</th>
              {/* <th>合計</th>
              <th>Extra</th>
              <th>MN Extra</th> */}
              <th>プロジェクト</th>
              <th>承認</th>
            </tr>
          </thead>
          <tbody>
            {days.map((a, index) => (
              <tr
                key={a.dailyId}
                className={a.hasUpdated ? "attendance-diff" : ""}
              >
                {/* <td
                  className={
                    a.isWeekday ? "attendance-weekday" : "attendance-holiday"
                  }
                >
                  {zerofill(2, index + 1)}
                </td> */}
                {a.isWeekday ? (
                  <TdWeekday>{zerofill(2, index + 1)}</TdWeekday>
                ) : (
                  <TdHoliday>{zerofill(2, index + 1)}</TdHoliday>
                )}
                <td className="tms-textfield tms-textfield--table">
                  <input
                    type="time"
                    value={a.start}
                    list="timelist-am"
                    onChange={e =>
                      updateDaily(a.dailyId, {
                        start: e.target.value
                      })
                    }
                  />
                </td>
                <td className="tms-textfield tms-textfield--table">
                  <input
                    type="time"
                    value={a.end}
                    list="timelist-pm"
                    onChange={e =>
                      updateDaily(a.dailyId, { end: e.target.value })
                    }
                  />
                </td>
                {/* プロジェクトが未設定の場合は表示なし */}
                <td>
                  {/* <SelectBox
                    value={a.project ? a.project.projectId : ""}
                    options={{
                      items: projects,
                      valueKey: "projectId",
                      labelKey: "name"
                    }}
                    onChange={$select =>
                      updateDaily(a.dailyId, {
                        projectId: $select.selectedOptions[0].value
                      })
                    }
                  // additionalClass={"tms-select--table"}
                  /> */}
                  select box
                </td>
                <td>
                  {a.hasConfirmed
                    ? MsgAttendance.HAS_SUBMITED
                    : MsgAttendance.NOT_SUBMITED}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }

  function onSaveClick(): void {
    saveAttendances();
  }

  function onSetDefaultClick(): void {
    setMonthlyWithDefaults(attendanceMonthly.days.map(d => d.dailyId));
  }

  function onNextMonthClick(): void {
    const { year: curYear, month: curMonth } = attendanceMonthly;
    const nextMonth = (curMonth % 12) + 1;
    const nextYear = nextMonth === 1 ? curYear + 1 : curYear;
    changeMonth(nextYear, nextMonth);
  }

  function onPrevMonthClick(): void {
    const { year: curYear, month: curMonth } = attendanceMonthly;
    const prevMonth = ((curMonth + 10) % 12) + 1;
    const prevYear = prevMonth === 12 ? curYear - 1 : curYear;
    changeMonth(prevYear, prevMonth);
  }

  function onSubmit(): void {
    console.log("[Attendance] Submitted");
  }
};

const TdHoliday = styled.td`
  color: rgba(255, 0, 0, 0.4);
  font-weight: ${BOLD};
`;

const TdWeekday = styled.td``;
