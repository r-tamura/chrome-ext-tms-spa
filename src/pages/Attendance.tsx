import * as React from "react";
import { connect } from "react-redux";
import { Helmet } from "react-helmet";
import { RouteComponentProps } from "react-router-dom";
import SelectBox from "~/components/SelectBox";
import { Button } from "~/components/Button";
import {
  RootState,
  getProjects,
  getAttendancesSelectedMonth,
  getAttendanceSettings
} from "~/modules";
import {
  fetchAttendancesIfNeeded,
  saveAttendancesIfNeeded,
  updateDaily,
  fetchSettings,
  setMonthlyWithDefaults,
  updateSettings,
  changeMonth
} from "~/modules/attendances";
import { zerofill } from "~/helpers/common";
import { AttendanceMonthlyView, AttendanceSettings, Project } from "~/types";
import { MsgAttendance } from "~/helpers/_const";

interface OwnProps extends RouteComponentProps<{}>, React.Props<{}> {}

interface IProps extends OwnProps {
  attendanceMonthly: AttendanceMonthlyView;
  attendanceSettings: AttendanceSettings;
  projects: Project[];
  fetchAttendancesIfNeeded: () => any;
  fetchSettings: () => any;
  saveAttendancesIfNeeded: () => any;
  updateDaily: (dailyId: string, patch: any) => any;
  setMonthlyWithDefaults: (dailyIds: string[]) => any;
  updateSettings: (patch: Partial<AttendanceSettings>) => any;
  changeMonth: (year: number, month: number) => any;
  submitApplication: (year: number, month: number) => any;
}

class AttendancePage extends React.Component<IProps, {}> {
  constructor(props: IProps) {
    super(props);
    this.props.fetchSettings();
    this.props.fetchAttendancesIfNeeded();
  }

  public render() {
    const { attendanceMonthly } = this.props;
    const { year, month, days } = attendanceMonthly;
    return (
      <main className="main">
        <Helmet>
          <title>Attendance | TMS</title>
        </Helmet>
        <div className="main-column">
          <h1>
            <Button id={"btn-prev-month"} onClick={this.onPrevMonthClick}>
              {"<"}
            </Button>
            {year} / {month}
            <Button id={"btn-next-month"} onClick={this.onNextMonthClick}>
              {">"}
            </Button>
          </h1>
          {this.renderMonthly(attendanceMonthly)}

          {/* 時間リスト */}
          <datalist id="timelist-am">
            <option>09:00</option>
            <option>09:30</option>
            <option>10:00</option>
            <option>10:30</option>
          </datalist>
          <datalist id="timelist-pm">
            <option>17:30</option>
            <option>18:00</option>
            <option>18:30</option>
            <option>19:00</option>
            <option>19:30</option>
            <option>19:30</option>
            <option>20:00</option>
            <option>20:30</option>
            <option>21:00</option>
            <option>21:30</option>
            <option>22:00</option>
            <option>22:30</option>
            <option>23:00</option>
            <option>23:30</option>
          </datalist>
        </div>
      </main>
    );
  }

  private renderMonthly(attendanceMonthly: AttendanceMonthlyView) {
    const { year, month, days, isFetching, hasApplied } = attendanceMonthly;

    if (isFetching) {
      // TODO: ロードインジケータを表示するように
      return <p>Now Loading...</p>;
    }

    if (days.length === 0) {
      return <p>勤怠データを取得できません</p>;
    }

    return (
      <div>
        <div className={"tms-btn-group"}>
          <Button id={"btn-set-default"} onClick={this.onSetDefaultClick}>
            set default
          </Button>
          <Button id={"btn-save"} onClick={this.onSaveClick}>
            Save
          </Button>
          <Button
            id={"btn-fetch"}
            onClick={this.props.fetchAttendancesIfNeeded}
          >
            Reload
          </Button>
          {hasApplied ? (
            <Button id={"btn-submit"} title={"上長申請中"} disabled>
              Submit
            </Button>
          ) : (
            <Button
              id={"btn-submit"}
              title={"上長申請"}
              onClick={this.onSubmit}
            >
              Submit
            </Button>
          )}
        </div>
        <div>
          <SelectBox
            value={this.props.attendanceSettings.projectId || ""}
            options={{
              items: this.props.projects,
              valueKey: "projectId",
              labelKey: "name"
            }}
            onChange={$select =>
              this.props.updateSettings({
                projectId: $select.selectedOptions[0].value
              })
            }
          />
        </div>
        <table className="tms-table tms-table--bordered">
          <thead>
            <tr>
              <th>日</th>
              <th>出勤</th>
              <th>退勤</th>
              {/* <th>合計</th>
              <th>Extra</th>
              <th>MN Extra</th> */}
              <th>プロジェクト</th>
              <th>承認</th>
            </tr>
          </thead>
          <tbody>
            {days.map((a, index) => (
              <tr
                key={a.dailyId}
                className={a.hasUpdated ? "attendance-diff" : ""}
              >
                <td
                  className={
                    a.isWeekday ? "attendance-weekday" : "attendance-holiday"
                  }
                >
                  {zerofill(2, index + 1)}
                </td>
                <td className="tms-textfield tms-textfield--table">
                  <input
                    type="time"
                    value={a.start}
                    list="timelist-am"
                    onChange={e =>
                      this.props.updateDaily(a.dailyId, {
                        start: e.target.value
                      })
                    }
                  />
                </td>
                <td className="tms-textfield tms-textfield--table">
                  <input
                    type="time"
                    value={a.end}
                    list="timelist-pm"
                    onChange={e =>
                      this.props.updateDaily(a.dailyId, { end: e.target.value })
                    }
                  />
                </td>
                {/* プロジェクトが未設定の場合は表示なし */}
                <td>
                  <SelectBox
                    value={a.project ? a.project.projectId : ""}
                    options={{
                      items: this.props.projects,
                      valueKey: "projectId",
                      labelKey: "name"
                    }}
                    onChange={$select =>
                      this.props.updateDaily(a.dailyId, {
                        projectId: $select.selectedOptions[0].value
                      })
                    }
                    additionalClass={"tms-select--table"}
                  />
                </td>
                <td>
                  {a.hasConfirmed
                    ? MsgAttendance.HAS_SUBMITED
                    : MsgAttendance.NOT_SUBMITED}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }

  private onSaveClick = (e: React.MouseEvent<HTMLButtonElement>): void => {
    this.props.saveAttendancesIfNeeded();
  };

  private onSetDefaultClick = (
    e: React.MouseEvent<HTMLButtonElement>
  ): void => {
    this.props.setMonthlyWithDefaults(
      this.props.attendanceMonthly.days.map(d => d.dailyId)
    );
  };

  private onNextMonthClick = (e: React.MouseEvent<HTMLButtonElement>): void => {
    const { year: curYear, month: curMonth } = this.props.attendanceMonthly;
    const nextMonth = (curMonth % 12) + 1;
    const nextYear = nextMonth === 1 ? curYear + 1 : curYear;
    this.props.changeMonth(nextYear, nextMonth);
  };

  private onPrevMonthClick = (e: React.MouseEvent<HTMLButtonElement>): void => {
    const { year: curYear, month: curMonth } = this.props.attendanceMonthly;
    const prevMonth = ((curMonth + 10) % 12) + 1;
    const prevYear = prevMonth === 12 ? curYear - 1 : curYear;
    this.props.changeMonth(prevYear, prevMonth);
  };

  private onSubmit = (e: React.MouseEvent<HTMLButtonElement>): void => {
    console.log("[Attendance] Submitted");
  };
}

const mapStateToProps = (state: RootState, onwProps: OwnProps) => {
  return {
    projects: getProjects(state),
    attendanceSettings: getAttendanceSettings(state),
    attendanceMonthly: getAttendancesSelectedMonth(state)
  };
};

export { AttendancePage };

export default connect(
  mapStateToProps,
  {
    fetchAttendancesIfNeeded,
    saveAttendancesIfNeeded,
    updateDaily,
    fetchSettings,
    setMonthlyWithDefaults,
    updateSettings,
    changeMonth
  }
)(AttendancePage);
