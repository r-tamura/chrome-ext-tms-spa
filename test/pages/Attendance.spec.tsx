import "jest"
import * as React from "react"
import renderer from "react-test-renderer"
import { shallow, mount } from "enzyme"
import Adapter from "enzyme-adapter-react-16"
import toJson from "enzyme-to-json"
import { AttendancePage } from "~/pages/Attendance"
import {
  getMockRouterProps,
  AttendanceSettingsModelBuilder,
  AttendanceViewModelBuilder,
  ProjectsBuilder,
} from "../__mocks__/"
import { AttendanceMonthlyView } from "~/types"

describe("<AttendancePage />", () => {

  it("shoud match existing snapshot", () => {
    const mockRoute = getMockRouterProps(null)
    const mockFunc = jest.fn()
    const page = renderer.create(
      <AttendancePage
        attendanceMonthly={AttendanceViewModelBuilder.build()}
        attendanceSettings={AttendanceSettingsModelBuilder.build()}
        projects={ProjectsBuilder.of().with2Projects().build()}
        fetchAttendancesIfNeeded={jest.fn()}
        fetchSettings={jest.fn()}
        saveAttendancesIfNeeded={jest.fn()}
        updateDaily={jest.fn()}
        setMonthlyWithDefaults={jest.fn()}
        updateSettings={jest.fn()}
        changeMonth={mockFunc}
        {...mockRoute}
      />
    ).toJSON()

    expect(page).toMatchSnapshot()
  })

  it("should be right next month", () => {
    const mockRoute = getMockRouterProps(null)
    const mockFunc = jest.fn()
    const w = shallow(
      <AttendancePage
        attendanceMonthly={AttendanceViewModelBuilder.build()}
        attendanceSettings={AttendanceSettingsModelBuilder.build()}
        projects={ProjectsBuilder.of().with2Projects().build()}
        fetchAttendancesIfNeeded={jest.fn()}
        fetchSettings={jest.fn()}
        saveAttendancesIfNeeded={jest.fn()}
        updateDaily={jest.fn()}
        setMonthlyWithDefaults={jest.fn()}
        updateSettings={jest.fn()}
        changeMonth={mockFunc}
        {...mockRoute}
      />
    )
    w.find("#btn-next-month").simulate("click")
    expect(mockFunc.mock.calls[0]).toEqual([2018, 2])

    w.find("#btn-prev-month").simulate("click")
    expect(mockFunc.mock.calls[1]).toEqual([2017, 12])
  })

  it("should make disable submit button", () => {
    const mockRoute = getMockRouterProps(null)
    const mockFunc = jest.fn()
    const w = mount(
      <AttendancePage
        attendanceMonthly={AttendanceViewModelBuilder.of().hasApplied().build()}
        attendanceSettings={AttendanceSettingsModelBuilder.build()}
        projects={ProjectsBuilder.of().with2Projects().build()}
        fetchAttendancesIfNeeded={jest.fn()}
        fetchSettings={jest.fn()}
        saveAttendancesIfNeeded={jest.fn()}
        updateDaily={jest.fn()}
        setMonthlyWithDefaults={jest.fn()}
        updateSettings={jest.fn()}
        changeMonth={mockFunc}
        {...mockRoute}
      />
    )
    const $submit = w.find("button#btn-submit")
    expect($submit.length).toBe(1)
    expect($submit.hasClass("disabled")).toBeTruthy()
  })
})
