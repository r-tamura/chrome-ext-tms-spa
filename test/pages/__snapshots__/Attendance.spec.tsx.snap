// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AttendancePage /> shoud match existing snapshot 1`] = `
<main
  className="main"
>
  <h1>
    <button
      id="btn-prev-month"
      onClick={[Function]}
    >
      &lt;
    </button>
    2018
     / 
    1
    <button
      id="btn-next-month"
      onClick={[Function]}
    >
      &gt;
    </button>
  </h1>
  <div>
    <div>
      <button
        id="btn-set-default"
        onClick={[Function]}
      >
        set default
      </button>
      <button
        id="btn-save"
        onClick={[Function]}
      >
        Save
      </button>
      <button
        id="btn-fetch"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
          }
        }
      >
        Reload
      </button>
    </div>
    <div>
      <div
        className="tms-select "
      >
        <select
          name=""
          onChange={[Function]}
          value="000001"
        >
          <option
            value="201108010001"
          >
            総務（営業）
          </option>
          <option
            value="201203010001"
          >
            社内教育（一般）
          </option>
        </select>
        <label />
      </div>
    </div>
    <table
      className="tms-table tms-table--bordered"
    >
      <thead>
        <tr>
          <th>
            日
          </th>
          <th>
            出勤
          </th>
          <th>
            退勤
          </th>
          <th>
            プロジェクト
          </th>
          <th>
            承認
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className=""
        >
          <td
            className="attendance-holiday"
          >
            01
          </td>
          <td
            className="tms-textfield tms-textfield--table"
          >
            <input
              list="timelist-am"
              onChange={[Function]}
              type="time"
              value="09:00"
            />
          </td>
          <td
            className="tms-textfield tms-textfield--table"
          >
            <input
              list="timelist-pm"
              onChange={[Function]}
              type="time"
              value="17:30"
            />
          </td>
          <td>
            <div
              className="tms-select tms-select--table"
            >
              <select
                name=""
                onChange={[Function]}
                value="000001"
              >
                <option
                  value="201108010001"
                >
                  総務（営業）
                </option>
                <option
                  value="201203010001"
                >
                  社内教育（一般）
                </option>
              </select>
              <label />
            </div>
          </td>
          <td>
            承認
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <datalist
    id="timelist-am"
  >
    <option>
      09:00
    </option>
    <option>
      09:30
    </option>
    <option>
      10:00
    </option>
    <option>
      10:30
    </option>
  </datalist>
  <datalist
    id="timelist-pm"
  >
    <option>
      17:30
    </option>
    <option>
      18:00
    </option>
    <option>
      18:30
    </option>
    <option>
      19:00
    </option>
    <option>
      19:30
    </option>
    <option>
      19:30
    </option>
    <option>
      20:00
    </option>
    <option>
      20:30
    </option>
    <option>
      21:00
    </option>
    <option>
      21:30
    </option>
    <option>
      22:00
    </option>
    <option>
      22:30
    </option>
    <option>
      23:00
    </option>
    <option>
      23:30
    </option>
  </datalist>
</main>
`;
